python/mxnet/gluon/_memmonger.py

```diff
@@ -0,0 +1,177 @@
+    plan_info: dict, optional
+        Used to hold plan information.
+    **kwargs:
+        The arguments to infer shape.
+    Returns
+    -------
+    alloc_sym: symbol
+        A symbol with force mirror tagged on the nodes for better allocation.
+    """
+    threshold = threshold << 20
+    sym = sym.__copy__()
+    internals = sym.get_internals()
+    input_shapes = {key: val.shape for key, val in inputs.items()}
+    input_shapes.update({key: val.shape for key, val in params.items()})
+    _, out_shapes, _ = internals.infer_shape(**input_shapes)
+    shape_dict = list(zip(internals.list_outputs(), out_shapes))
+    total_size = 0
+    local_size = 0
+    save_size = 0
+    max_size = 0
+    last_sb = None
+    last_local = 0
+    period = 1
+    last_stage = ''
+    stage_decision = ''
+
+    for idx, item in enumerate(shape_dict):
+        sb = internals[idx]
+        name, shape = item
+        if name in input_shapes:
+            continue
+        total_size += prod(shape) * 4
+        local_size += prod(shape) * 4
+        sb._set_attr(force_mirroring='True')
+
+        if sb.attr('mirror_stage') is not None:
+            stage = sb.attr('mirror_stage')
+            if stage == 'True' or stage != last_stage:
+                if local_size > threshold:
+                    save_size += prod(shape) * 4
+                    max_size = max(max_size, local_size)
+                    local_size = 0
+                    stage_decision = 'False'
+                    sb._set_attr(force_mirroring=stage_decision)
+                else:
+                    stage_decision = 'True'
+                    pass
+                last_stage = stage
+            elif stage == last_stage and stage_decision == 'False':
+                save_size += prod(shape) * 4
+                sb._set_attr(force_mirroring=stage_decision)
+
+    if plan_info is not None:
+        plan_info['max_size'] = max_size
+        plan_info['save_size'] = save_size
+    return sym
+
+
+def get_cost(sym, inputs, params):
+    """Get the cost of the current symbolic plan by running bind on CPU.
+    sym : Symbolic Variable
+    """
+    grad_reqs = {}
+    type_dict = {}
+    shape_dict = {}
+    for key, val in inputs.items():
+        type_dict[key] = val.dtype
+        shape_dict[key] = val.shape
+        if val.grad is None:
+            grad_reqs[key] = 'null'
+        else:
+            grad_reqs[key] = 'write'
+
+    for key, val in params.items():
+        type_dict[key] = val.dtype
+        shape_dict[key] = val.shape
+        grad_reqs[key] = val.grad_req
+
+    texec = sym.simple_bind(ctx=cpu(),
+                            grad_req=grad_reqs,
+                            type_dict=type_dict,
+                            **shape_dict)
+    return int(texec.debug_str().split('\n')[-3].split()[1])
+
+
+def search_plan(sym, inputs, params, ntrials=6):
+    """Quickly heurestic search over possible plans to find good memory plan.
```

Comment:
Could you describe the heuristic* used to find the memory plan? Could you also describe for the users according to which metric do we consider a memory plan good?